// userop/dist/v06/account/common/simpleAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.base = void 0;
const simpleAccount_1 = require("./abi/simpleAccount");
const hooks_1 = require("../hooks");
const FACTORY_ADDRESS = "0x9406Cc6185a346906296840746125a0E44976454";
function base(ethClient, eoa, account) {
    return {
        accountAbi: simpleAccount_1.AccountAbi,
        factoryAbi: simpleAccount_1.FactoryAbi,
        factoryAddress: FACTORY_ADDRESS,
        ethClient,
        async setFactoryData(salt, encoder) {
            if (account !== undefined) {
                return encoder("createAccount", [account.address, salt]);
            }
            else if ("getAddresses" in eoa) {
                return encoder("createAccount", [(await eoa.getAddresses())[0], salt]);
            }
            else {
                return encoder("createAccount", [
                    (await eoa.getAddress()),
                    salt,
                ]);
            }
        },
        requestSignature: "getAddresses" in eoa
            ? account !== undefined
                ? hooks_1.RequestSignature.withViemWalletClient(eoa, account)
                : hooks_1.RequestSignature.withViemWalletClient(eoa)
            : hooks_1.RequestSignature.withEthersSigner(eoa),
    };
}
exports.base = base;
