// userop/dist/v06/account/hooks/requestPaymaster/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withCommon = void 0;
const viem_1 = require("viem");
const stackupV1Paymaster = (params) => {
    const client = (0, viem_1.createPublicClient)({ transport: (0, viem_1.http)(params.rpcUrl) });
    return async (userop, entryPoint) => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { rpcUrl, ...ctx } = params;
        const pm = (await client.transport.request({
            method: "pm_sponsorUserOperation",
            params: [userop, entryPoint, ctx],
        }));
        return {
            paymasterAndData: pm.paymasterAndData,
            callGasLimit: BigInt(pm.callGasLimit),
            verificationGasLimit: BigInt(pm.verificationGasLimit),
            preVerificationGas: BigInt(pm.preVerificationGas),
        };
    };
};
const withCommon = (opts) => {
    switch (opts.variant) {
        case "stackupV1":
            return stackupV1Paymaster(opts.parameters);
    }
};
exports.withCommon = withCommon;
