// userop/dist/v06/account/hooks/requestGasValues/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withEthClient = void 0;
const ethers_1 = require("ethers");
const entryPoint_1 = require("../../../entryPoint");
const withEthClient = (ethClient) => {
    if (ethClient instanceof ethers_1.JsonRpcProvider) {
        return async (userop, entryPoint, stateOverrideSet) => {
            const est = (await ethClient.send("eth_estimateUserOperationGas", stateOverrideSet !== undefined
                ? [(0, entryPoint_1.toRawUserOperation)(userop), entryPoint, stateOverrideSet]
                : [(0, entryPoint_1.toRawUserOperation)(userop), entryPoint]));
            return {
                preVerificationGas: BigInt(est.preVerificationGas),
                verificationGasLimit: BigInt(est.verificationGasLimit),
                callGasLimit: BigInt(est.callGasLimit),
            };
        };
    }
    return async (userop, entryPoint, stateOverrideSet) => {
        const est = (await ethClient.transport.request({
            method: "eth_estimateUserOperationGas",
            params: stateOverrideSet !== undefined
                ? [userop, entryPoint, stateOverrideSet]
                : [userop, entryPoint],
        }));
        return {
            preVerificationGas: BigInt(est.preVerificationGas),
            verificationGasLimit: BigInt(est.verificationGasLimit),
            callGasLimit: BigInt(est.callGasLimit),
        };
    };
};
exports.withEthClient = withEthClient;
